### ISSUE TITLE
Timelines is locked for very long time

### ISSUE THREAD

**Issue by knizhnik** [zenithdb/zenith#1087]
With write amplification test (https://github.com/zenithdb/zenith/issues/1076)
I constantly get `Timed out while waiting for WAL record` error after few hours of work.
I suspected that it was caused by locks hold by GC and which prevents other threads to do their work (i.e. sending responses to  walproposer).
And looks like this hypothesis is proved: I got the following messages before timeout happen:
```
2022-01-05T20:34:30.757113Z  INFO WAL receiver{timeline=36993e17c766caba2767fceb3291f101 tenant=85a9083d299a4d6cb7ba80
8af60da97d}: get_timeline_state wait for timelines lock 414.925164583s
2022-01-05T20:34:30.757166Z  INFO callmemaybe{timeline=36993e17c766caba2767fceb3291f101 tenant=85a9083d299a4d6cb7ba808
af60da97d}: get_timeline lock 411.211896242s
connection error: error communicating with the server: Broken pipe (os error 32)
```

So `get_timeline_state` waits for 414 seconds for `self.timelines` lock and definitely walproposer has decided that peer is not alive and drop connection.

I will try to replace std::mutex wtih parking_lot, but may be there is better solution of the problem?

**Comment by knizhnik** [comment]
Just replacing std::mutex with parking_lot::mutex in layered_repository.rs doesn't help.
So the problem is not caused by lock starvation...

**Comment by knizhnik** [comment]
Finally I produced core dump and find out who is locking `timelines` mutex.
Actually this problem is well known and I have reported it long time ago: GC is now performed with `timelines` mutex lock. In my test explicitly GC is called with  `checkpoint_before_gc=true` which makes times of GC iteration (and so `timelines` lock) inconceivably long. 
But even without GC holding `timelines` lock for the whole time of GC operation IMHO is not acceptable.

**Comment by knizhnik** [comment]
So I think that under the lock we should construct list of timelines to be garbage collected and then release `timelines` lock and perform GC without holding this lock.

**Comment by hlinnaka** [comment]
Was this fixed by commit f70a5cad61e756f3185522ff14daf0de9a117ec8?

**Comment by knizhnik** [comment]
Yes, this problem is fixed. But it introduce new problems (race conditions) which has to be addressed (#1118). There is one fix which is not yet committed in main: hold checkpoint_cs in GC (to prevent race condition between GC and checkpointer). But I am not sure if it is really good idea.

**Comment by hlinnaka** [comment]
I think we're all done here, closing.
